horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/RunServiceImpl.java

```diff
@@ -1004,4 +1011,216 @@ public DatasetList listDatasetsBySchema(String uri, Integer limit, Integer page,
+                        .setParameter(1, run.id).setParameter(2, schemaid).setParameter(3, transformerId)
+                        .unwrap(NativeQuery.class)
+                        .addScalar("name", TextType.INSTANCE)
+                        .addScalar("value", JsonNodeBinaryType.INSTANCE)
+                        .getResultList() );
+                     root = convert(root, extractorValues);
+                  } else {
+                     extractorValues = unchecked(em.createNamedQuery(QUERY_2ND_LEVEL_BY_RUNID_TRANSFORMERID_SCHEMA_ID)
+                        .setParameter(1, run.id).setParameter(2, run.id).setParameter(3, schemaid).setParameter(4, transformerId)
+                        .unwrap(NativeQuery.class)
+                        .addScalar("name", TextType.INSTANCE)
+                        .addScalar("value", JsonNodeBinaryType.INSTANCE)
+                        .getResultList() );
+                     root = convert(root, extractorValues);
+                  }
+               }
+               StringBuilder code = new StringBuilder();
+               String obj = root.toString();
+               code.append("const __obj = ").append(obj).append(" ;\n");
+               code.append("const __func = ").append(t.function).append(";\n");
+               code.append("__func(__obj);\n");
+               ByteArrayOutputStream out = new ByteArrayOutputStream();
+               try (org.graalvm.polyglot.Context context = org.graalvm.polyglot.Context.newBuilder("js").out(out).err(out).build()) {
+                  context.enter();
+                  try {
+                     Value value = context.eval("js", code.toString());
+                     result = Util.convertToJson(value);
+                  } catch (PolyglotException e) {
+                     log.errorf("Failed to execute transformation for run %d and test %d and cause %s", run.id, test.id, e);
+                     if (result != null) {
+                        logTransformationMessage(test, run, TransformationLog.ERROR, "transformation "+t.name+" function", result);
+                        result = instance.objectNode();
+                     }
+                  } finally {
+                     context.leave();
+                     if (out.size() > 0) {
+                        logTransformationMessage(test, run, TransformationLog.ERROR, "transformation "+t.name+" function", result);
+                     }
+                  }
+               }
+               if (t.targetSchemaUri != null && result != null) {
+                  if (result.isObject()) {
+                     ((ObjectNode)result).put("$schema", t.targetSchemaUri);
+                  } else if (result.isArray()) {
+                     ArrayNode array = (ArrayNode)result;
+                     ArrayNode replacement = instance.arrayNode();
+                     for ( JsonNode node : array ) {
+                        ObjectNode n = instance.objectNode();
```

Comment:
> If node is an object there's no reason to nest it further, your code would convert [ { "foo": "bar" } ] into [ { "$schema": "...", "value": { "foo": "bar" } } ] while it can be simply [ { "foo": "bar", "$schema": "..." } ].

 Don't know where the `value` field name came from. It will be removed.

> Moreover you're overwriting if there already is a $schema (can be determined dynamically)

 Ok, I'll only set it if not present.